File: PHP/680. Valid Palindrome II.php
<?php

/**
 * @noinspection PhpIllegalPsrClassPathInspection
 * @noinspection PhpMultipleClassDeclarationsInspection
 * @noinspection PhpUnused
 */
class Solution
{
    public function validPalindrome(string $s): bool
    {
        $length = strlen($s);
        if ($length <= 1) {
            return true;
        }

        $left = 0;
        $right = $length - 1;
        while ($left < $right) {
            if ($s[$left] !== $s[$right]) {
                return $this->palindrome($s, $left, $right - 1) || $this->palindrome($s, $left + 1, $right);
            }

            $left++;
            $right--;
        }

        return true;
    }

    private function palindrome(string $s, int $left, int $right): bool
    {
        while ($left < $right) {
            if ($s[$left] !== $s[$right]) {
                return false;
            }

            $left++;
            $right--;
        }

        return true;
    }
}

File: Go/680. Valid Palindrome II.go
package main

//goland:noinspection GoUnusedFunction
func validPalindrome(s string) bool {
	length := len(s)

	if length <= 1 {
		return true
	}

	left := 0
	right := length - 1

	for left < right {
		if s[left] != s[right] {
			return palindrome(s, left+1, right) || palindrome(s, left, right-1)
		}

		left++
		right--
	}

	return true
}

func palindrome(s string, left int, right int) bool {
	for left < right {
		if s[left] != s[right] {
			return false
		}

		left++
		right--
	}

	return true
}
